src/many-identity-dsa/src/impls/ed25519.rs

```diff
@@ -0,0 +1,423 @@
+                    .ok_or_else(|| ManyError::unknown("Could not get EdDSA X parameter"))?;
+                Ok(Some(eddsa_cose_key(x, None)))
+            } else {
+                Err(ManyError::unknown("Key doesn't have a public key"))
+            }
+        }
+        _ => Ok(None),
+    }
+}
+
+/// Verify that key is valid Ed25519.
+fn check_key(cose_key: &CoseKey, sign: bool, verify: bool) -> Result<(), ManyError> {
+    if sign
+        && !cose_key.key_ops.contains(&coset::KeyOperation::Assigned(
+            coset::iana::KeyOperation::Sign,
+        ))
+    {
+        return Err(ManyError::unknown("Key cannot sign"));
+    }
+    if verify
+        && !cose_key.key_ops.contains(&coset::KeyOperation::Assigned(
+            coset::iana::KeyOperation::Verify,
+        ))
+    {
+        return Err(ManyError::unknown("Key cannot verify"));
+    }
+
+    if cose_key.kty != coset::KeyType::Assigned(coset::iana::KeyType::OKP) {
+        return Err(ManyError::unknown(format!(
+            "Wrong key type: {:?}",
+            cose_key.kty
+        )));
+    }
+    if cose_key.alg != Some(coset::Algorithm::Assigned(coset::iana::Algorithm::EdDSA)) {
+        return Err(ManyError::unknown(format!(
+            "Wrong key algorihm: {:?}",
+            cose_key.alg
+        )));
+    }
+
+    if cose_key
+        .params
+        .iter()
+        .find(|(k, _v)| k == &Label::Int(OkpKeyParameter::Crv.to_i64()))
+        .map(|(_k, v)| v)
+        .ok_or_else(|| ManyError::unknown("Crv parameter not found."))?
+        .as_integer()
+        .ok_or_else(|| ManyError::unknown("Crv parameter not found."))?
+        != Integer::from(coset::iana::EllipticCurve::Ed25519.to_i64())
+    {
+        Err(ManyError::unknown("Curve unsupported. Expected Ed25519"))
+    } else {
+        Ok(())
+    }
+}
+
+fn key_pair(cose_key: &CoseKey) -> Result<Keypair, ManyError> {
+    check_key(cose_key, true, false)?;
+
+    let mut maybe_x = None;
+    let mut maybe_d = None;
+    for (k, v) in cose_key.params.iter() {
+        if k == &Label::Int(OkpKeyParameter::X.to_i64()) {
+            maybe_x = Some(v);
+        } else if k == &Label::Int(OkpKeyParameter::D.to_i64()) {
+            maybe_d = Some(v);
+        }
+    }
+
+    let x = maybe_x
+        .ok_or_else(|| ManyError::unknown("Could not find the X parameter in key"))?
+        .as_bytes()
+        .ok_or_else(|| ManyError::unknown("Could not convert the X parameter to bytes"))?
+        .as_slice();
+    let d = maybe_d
+        .ok_or_else(|| ManyError::unknown("Could not find the D parameter in key"))?
+        .as_bytes()
+        .ok_or_else(|| ManyError::unknown("Could not convert the D parameter to bytes"))?
+        .as_slice();
+
+    Keypair::from_bytes(&vec![d, x].concat())
+        .map_err(|e| ManyError::unknown(format!("Invalid Ed25519 keypair from bytes: {e}")))
+}
+
+#[derive(Clone)]
+struct Ed25519IdentityInner {
+    address: Address,
+    public_key: CoseKey,
+    key_pair: Arc<Keypair>,
```

Comment:
Why is `Arc` needed?

Reply:
`Keypair` does not have `Clone`.

Reply:
I removed `Clone` from all keys. We can reintroduce it later if we need to.